### DISCUSSION TITLE
Possible huge write performance boost

### DISCUSSION THREAD

**Discussion by quolpr** [rhashimoto/wa-sqlite#57] [Ideas]
I was discovering why absurd-sql is very fast even on large datasets write, and why wa-sqlite doesn't have such performance (when table has 100,000+ records). And I discovered something interesting — absurd-sql makes write only on transaction finish. During the transaction, it makes reads from IDB, but writes are happening in-memory. Then it dumps all the data to IDB on transaction finish https://github.com/jlongster/absurd-sql/blob/master/src/sqlite-file.js#L374

**Comment by quolpr** [comment]
When I make insert of a lot of data, I am getting a lot of reeds and the write requests:

<img width="946" alt="image" src="https://user-images.githubusercontent.com/7958527/178099573-8d0c11f4-36c8-4088-8445-4b0dbb0e2b9e.png">

If we could postpone the writing to the final step(on transaction commit), then it should improve performance. I am using IDBBatchAtomicVFS

**Reply by quolpr** [reply]
Actually, I tried to add some cache on data write during transaction, and if the block will be read again from the same offset — it will have a cache hit. I also reset cache on `SQLITE_FCNTL_SYNC`. The write performance improved dramatically 🤔

**Comment by rhashimoto** [comment]
Hmm. I'm not sure why buffering the writes should speed things up, provided that wa-sqlite is using a single IDB transaction for a SQLite transaction (you can log IDB transactions by uncommenting [this line](https://github.com/rhashimoto/wa-sqlite/blob/29a38da3dc001f8e6844837f6f3dffcdda2cdb10/src/examples/IDBContext.js#L7) to check if it is succeeding in minimizing transactions). The writes need to go to IDB eventually, so I don't see an obvious advantage in deferring them. The only thing I can think of is that if a lot of blocks are written more than once in the transaction then overwrite in memory might be a lot faster than overwrite in IDB, but I wouldn't be certain of that because I would hope that expense would be mostly off-thread.

If a lot of blocks are being read after being written, then yes, that's going to be a performance hit. I would see whether [increasing the SQLite cache size](https://www.sqlite.org/pragma.html#pragma_cache_size) helps to avoid that. I would give that a try before inserting another cache at the VFS level.

**Reply by quolpr** [reply]
> If a lot of blocks are being read after being written, then yes, that's going to be a performance hit. I would see whether [increasing the SQLite cache size](https://www.sqlite.org/pragma.html#pragma_cache_size) helps to avoid that. I would give that a try before inserting another cache at the VFS level.

I made block size 32kb and cache_size to -5000, but it is still suffering if whole database can't be stored at memmory.

I made this for xRead:

```
      const fromCache = this.cache.get(-iOffset);
      if (fromCache && fromCache.length === pData.size) {
        pData.value.set(fromCache);

        return VFS.SQLITE_OK;
      }
```

And this for xWrite:

```
      if (iOffset !== 0) {
        this.cache.set(-iOffset, block.data);
      }
```

You can check this commit — https://github.com/trong-orm/trong-orm/commit/e1da525157d14b6125c22f249e5c16998bc55d77

The write performance now even better than absurd-sql at some cases  🤔

**Reply by rhashimoto** [reply]
If I'm reading your code correctly, you don't limit your cache size. I'd be interested in how big it gets. Then compare the performance with an unchanged VFS with cache size increased to that number plus 2000 (assuming that's more than the 5000 you already tried). I think that would be closer to apples-to-apples memory usage. It's still possible your modified VFS would be faster since the SQLite cache contains both read and written blocks.

I had caches in some earlier VFS classes, but eventually I did away with them. I didn't like that blocks were duplicated if the SQLite cache was active, and disabling the SQLite cache isn't advisable because it is the fastest. I also capped my cache size because I wanted to allow very large transactions, but then I had to tune that cache size _and_ the SQLite cache size, plus it was more code and tracking down cache bugs is not a lot of fun. So now I think really hard about whether I want to implement another cache and so far the answer has been no. Just something to consider even if there is a performance win.

I guess it shouldn't be surprising that IDBBatchAtomicVFS can be competitive with absurd-sql on pure write performance (though I wasn't expecting that). `IDBBatchAtomicVFS.xWrite` is synchronous (unlike `xRead`), so the Asyncify overhead should be small, i.e. it doesn't have to do any unwinding or rewinding.

**Reply by quolpr** [reply]
Yep, I don't limit. That is downside, and memory problems could be here.

Also, thank you so much for detailed response! It helps me a lot 🙂 And thank you for the all the work you did, it's an excellent job.

**Reply by quolpr** [reply]
Some update: actually increasing cache size to -20,000 did the job — the write performance is the same as with the cache. You were right!

**Reply by quolpr** [reply]
And some other update: but actually, after some more inserts, the write performance begin to degrade (the DB become larger). While with cache — it remains constant, or degrading not so fast 🤔

**Comment by quolpr** [comment]
@rhashimoto finally I was able to achieve the same write performance as absurd-sql has. I am not sure why, but your IDBContext/Store has a bottleneck somewhere. So instead of:

```
await db.run("readwrite", async ({ blocks }) => {
  await Promise.all(
    Array.from(this.blockToWrite.keys()).map(async (k) => {
      const toWrite = toWriteBlocks.get(k)!;

      await blocks.put(toWrite);
    })
  );
});
```

I put:

```
await new Promise<void>((resolve, reject) => {
  // @ts-expect-error lib dom misses third argument
  const tx = idb.transaction("blocks", "readwrite", {
    durability: "relaxed",
  });

  const blocksStore = tx.objectStore("blocks");

  for (const k of toWriteBlocks.keys()) {
    const toWrite = toWriteBlocks.get(k)!;

    blocksStore.put(toWrite);
  }

  tx.addEventListener("complete", () => {
    resolve();
  });
  tx.addEventListener("abort", reject);
  tx.addEventListener("error", reject);
});
```

And this code run double faster than the first one 🤔

**Reply by rhashimoto** [reply]
It's always possible there's a bug in IDBContext, but in your first code block there is no need to await each `put()` and I would expect that to leak performance. The equivalent code should look more like this (assuming the IDBContext instance is configured for the same durability):

```javascript
db.run("readwrite", ({ blocks }) => {
  // assuming toWriteBlocks is a Map...
  for (const block of toWriteBlocks.values()) {
    blocks.put(block);
  }
});
await db.sync();
```

**Reply by rhashimoto** [reply]
Also, if you are going to journal only to memory and defer all writes to the end of a transaction then you don't actually need the batch atomic writes and versioned blocks in IDBBatchAtomicVFS. Removing all that would add a little more performance. For example, versioned blocks uses an IndexedDB index so every time blocks are added or removed the index is also updated.

If that is your path then you might be better off using [IDBMinimalVFS](https://github.com/rhashimoto/wa-sqlite/discussions/46) as your starting point (instead of IDBBatchAtomicVFS). IDBMinimalVFS doesn't support changing the page size of a database so if you want to support that then you might need to adapt [IDBBatchAtomicVFS.#reblockIfNeeded()](https://github.com/rhashimoto/wa-sqlite/blob/9962609ed199d6d53899f3e5f1c54369fa9aee58/src/examples/IDBBatchAtomicVFS.js#L530-L616), but I still think that it would be easier to augment something simple than to pare down something complicated.

**Reply by quolpr** [reply]
> The equivalent code should look more like this (assuming the IDBContext instance is configured for the same durability)

It works with the same performance 🙂 Thanks!

> If that is your path then you might be better off using https://github.com/rhashimoto/wa-sqlite/discussions/46 as your starting point (instead of IDBBatchAtomicVFS).

Yep, I start to think the same. Thanks!

**Reply by quolpr** [reply]
Here is the final implementation (I hope so 😄 But reblock is missing):

https://github.com/trong-orm/trong-orm/blob/8dc69d9ac33208fe6f7837a489e79bf4def9c327/packages/wa-sqlite-web-backend/src/IDBMinimalVFS.ts

I also tested your code again:

```
db.run("readwrite", ({ blocks }) => {
  // assuming toWriteBlocks is a Map...
  for (const block of toWriteBlocks.values()) {
    blocks.put(block);
  }
});
await db.sync();
```

And it seems `await db.sync()` don't await all the transaction to finish (maybe only the last one? Not sure, needs more check)

I put the code with raw IDB call for now. I think I will refactor the code, so it will not use `IDBContext`. Not only that, but I think it is possible to make IDBContext in that way it will more clear & type safe  🤔 

And yep, writing performance now is the same as absurd has 🙂

**Reply by rhashimoto** [reply]
Are you comparing with absurd-sql with [your proposed commit](https://github.com/jlongster/absurd-sql/pull/56)? Because I think your change doesn't really do anything except make it slower. Transaction requests are already queued and executed sequentially; there is no need to issue them in their own task (which is what your code does). See the [spec](https://www.w3.org/TR/IndexedDB-2/#transaction-lifetime-concept):

> The implementation must allow requests to be placed against the transaction whenever the active flag is set. This is the case even if the transaction has not yet been started. Until the transaction is started the implementation must not execute these requests; however, the implementation must keep track of the requests and their order.
> [...]
> Unless otherwise defined, requests must be executed in the order in which they were made against the transaction. Likewise, their results must be returned in the order the requests were placed against a specific transaction.

If you aren't comparing with the stock absurd-sql then it won't be a fair race (for absurd-sql).

**Reply by quolpr** [reply]
No, I think my commit is right. The problem is that if you will make read after write without awaiting writing to finish, then the next SELECT query will be delayed for the time writing is finishing. 

Here is the video, where I comment and uncomment await of insert, and then to select query. If I do await, then next select query if fast. If I not, the next select query is slow (cause it awaits the IDB to finish writing):

https://user-images.githubusercontent.com/7958527/178580556-120c049c-45cd-417c-91db-806894c960c3.mp4

So do get actual write performance, you should await all IDB write requests to finish

**Reply by quolpr** [reply]
Even I was not able to achieve such writing performance as abusrd-sql has on their benchmark with raw ArrayBuffer writing into IDB (for the same ArrayBuffer size and the same count of such blocks, and it was running in one readwrite transaction), that is already very suspicious

**Reply by quolpr** [reply]
> Transaction requests are already queued and executed sequentially; there is no need to issue them in their own task (which is what your code does)

Not awaiting writing to finish actually create IDB tasks too, and blocks all future reads (what you can see on the demo I have attached)

I hope I am right 🤔

**Reply by quolpr** [reply]
And here is the example with `Promise.all`. The performance is around the same. I also want to make sure you, that there is no new transactions spawn during write — everything happen in one write transaction.


https://user-images.githubusercontent.com/7958527/178582720-7fcb5763-9ffb-4928-8e27-930576177b06.mp4

**Reply by rhashimoto** [reply]
This is not the way to time things. You are making the code slower in order to time it:

```javascript
console.time('bulkSet');
for (let item of items) {
  await this.set(item);
}
console.timeEnd('bulkSet');
```

Your attempt with `Promise.all()` is still bad, but slightly better:

```javascript
console.time('bulkSet');
const res = await Promise.all(
  items.map(async (it) => {
    await this.set(it);
  })
);
console.timeEnd('bulkSet');
```

Do you understand why this is better? You're still slowing things down unnecessarily, but at least you're not waiting for the previous put before submitting the next put. The code in your original commit takes N tasks to request N puts, and by "task" I mean an [event loop macrotask](https://javascript.info/event-loop).

Here is a much better way to do this timing:

```javascript
console.time('bulkSet');
Promise.all(items.map(it => this.set(it)).finally(() => {
  console.timeEnd('bulkSet');
});
```

This measures the same thing but doesn't negatively affect the performance of the code being measured.

In addition, your PR message indicates that you think this change may fix a database corruption bug, but I don't see how this is plausible.

**Reply by quolpr** [reply]
So, from your point of view, the next read query(like SELECT) that will be waiting for write(cause IDB still wait for the all puts to do, doesn't matter if you do await or not in terms of next IDB read) to finish is a good thing? 🤔 

I mean, it is not fair, that insert is fast, but the next query will be much slower cause actual insert is still not finished

**Reply by rhashimoto** [reply]
It is fine to be curious about where and how cycles are being spent. It is rarely if ever useful to significantly change _how_ they are being spent in order to measure them. If you do want to allocate times to particular things then you should be doing it in a much less invasive way (I tried to show you such a way). And even if invasive means were the only way to do the measurement (which is _not_ the case here) why would you want to commit that to production code?

And yes, I don't mind that the queueing of IDB operations is done off-thread instead of in my code. In fact, that's exactly where I want it to be. When IndexedDB throughput is my critical path, I want to keep that pipeline full as much as I can, and that means not going back to Javascript for every operation.

Here's a [microbenchmark illustration](https://jsfiddle.net/rhashimoto/mktvo9ds/43/) of what idling the pipeline can cost you. Each button launches a large number of IndexedDB puts on a newly created database, one awaits puts and one does not. The time to complete the entire transaction is reported. Using Chrome on my (old and slow) machine, with await takes ~21 seconds and without await takes ~11 seconds. This is not just a matter of shifting time assignments from one put to another - it has a negative impact on overall performance.

There might be a time when I would want to study which statement and what operations in a multi-statement SQL transaction is the most expensive and why. But I doubt that injecting awaits will be the appropriate way to investigate it, and I certainly wouldn't commit added awaits to production unless they fix an actual bug.

**Reply by quolpr** [reply]
>  If you do want to allocate times to particular things then you should be doing it in a much less invasive way (I tried to show you such a way).

Who says which way should I do the things?

> And even if invasive means were the only way to do the measurement (which is not the case here) why would you want to commit that to production code?

I open to any discussion, why do you think that I want to commit strait forward to the production code? I agree, that I should put `Promise.all` to the PR, that is fair point. But why do you say, that I force pushing it to the production code?

If you main that thing, that I made a fork of absurd-sql — you still have the choice to not use it. I don't push anybody to use my fork or to agree with my discussions.

> And yes, I don't mind that the queueing of IDB operations is done off-thread instead of in my code. In fact, that's exactly where I want it to be. When IndexedDB throughput is my critical path, I want to keep that pipeline full as much as I can, and that means not going back to Javascript for every operation.

Why do you think, that the author absurd-sql did the same thing as you? The absurd-sql code part looks more like a mistake, cause in method declaration `async` is present, but no `await` in method body.

And if it was conceived, why the author didn't mention it in the article? Like, hey, write operations are faster because they are actually delayed. And to make fair comparison, you should actually do write+read benchmark in the same row.

**Reply by quolpr** [reply]
> Your attempt with Promise.all() is still bad, but slightly better:

But why if I put `Promise.all` on your benchmark it shows the same number, as without awaiting? https://jsfiddle.net/adsoyqr1/3/

And it is not `slightly`, it actually the same

<img width="319" alt="image" src="https://user-images.githubusercontent.com/7958527/178829244-fe05a2eb-3d0b-43d0-aaed-9e1d7f78f5c0.png">

**Reply by rhashimoto** [reply]
[This is your pull request](https://github.com/jlongster/absurd-sql/pull/56) to the master branch of absurd-sql, right? That's what I'm talking about. It will make absurd-sql slower overall. I don't see the point of comparing with a baseline that you have made slower, which is why I asked what you were comparing with.

It looked like you thought you were also fixing a bug, and if that was a valid fix then you would be perfectly justified in comparing with the fix added. And so I tried to show you that your change could not fix that database corruption bug.

I'm not telling you what to do in your own tree. I'm telling you that your PR is flawed, which I think you now realize, and using a version of absurd-sql with that change in it to make comparisons (which I don't know if you were doing or not, hence my question) would not be valid.

**Reply by rhashimoto** [reply]
> But why if I put Promise.all on your benchmark it shows the same number, as without awaiting?

Yes, I said this was better, and you're correct that the microbenchmark shows no time penalty. But I still don't think this code is good. The reason the await doesn't hurt here is because there is nothing left to go into the IndexedDB pipeline before the transaction commits and everything goes away. Here's why I think this is still the wrong way to go:

1. In a real application you could still be blocking new operations from going into the IndexedDB pipeline. If the next thing the application does is open a new transaction and start creating requests on it, which is actually quite likely, await will delay that from happening. You have cleaned up the pipeline bubbles between operations but you're still risking bubbles between transactions that harm overall performance.

2. If performance is important to you, in general you need a pretty good reason (e.g. correctness) to block the critical path. Using benchmarking as one of those reasons seems so counter-productive.  [Code smell](https://en.wikipedia.org/wiki/Code_smell) is a subjective criticism, I admit, but in my opinion it applies here (and that isn't meant pejoratively, or to deny the stench in much of my own code).

3. There are workable alternatives. You can use the same IndexedDB event hooks to monitor progress, just without using await on the critical path.

**Reply by quolpr** [reply]
> https://github.com/jlongster/absurd-sql/pull/56 to the master branch of absurd-sql, right? That's what I'm talking about. It will make absurd-sql slower overall. I don't see the point of comparing with a baseline that you have made slower, which is why I asked what you were comparing with.

I am comparing just the real write performance, nothing more.

> It looked like you thought you were also fixing a bug, and if that was a valid fix then you would be perfectly justified in comparing with the fix added. And so I tried to show you that your change could not fix that database corruption bug.

Yes, you are right. I was wrong. But still, the inserts at `absurd-sql` are not measured right.

> In a real application you could still be blocking new operations from going into the IndexedDB pipeline. If the next thing the application does is open a new transaction and start creating requests on it, which is actually quite likely, await will delay that from happening. You have cleaned up the pipeline bubbles between operations but you're still risking bubbles between transactions that harm overall performance.

I agree for `wa-sqlite`, but not for `absurd-sql` case. When I put `Promise.all` on write, the next query still will be in IDB transaction, cause transaction almost never complete because worker uses Atomic awaiting (so the next call in event loop still will be an IDB call).

> If performance is important to you, in general you need a pretty good reason (e.g. correctness) to block the critical path. Using benchmarking as one of those reasons seems so counter-productive. [Code smell](https://en.wikipedia.org/wiki/Code_smell) is a subjective criticism, I admit, but in my opinion it applies here (and that isn't meant pejoratively, or to deny the stench in much of my own code).

I didn't say anything about code smell. And I didn't say that we should fix this cause of code smell. I just point, that James could just miss the await phase of `bulkWrite`
